domain/application/state/application.go

```diff
@@ -2991,29 +2992,396 @@ WHERE application_uuid = $applicationID.uuid;
+					return errors.Capture(err)
+				}
+			}
+		}
+
+		return errors.Capture(
+			tx.Query(ctx, insertAppConstraintsStmt, setApplicationConstraint{
+				ApplicationUUID: appID.String(),
+				ConstraintUUID:  cUUIDStr,
+			}).Run(),
+		)
+	})
+
+}
+
+// decodeConstraints flattens and maps the list of rows of applicatioConstraint
+// to get a single constraints.Value. The flattening is needed because of the
+// spaces, tags and zones constraints which are slices. We can safely assume
+// that the non-slice values are repeated on every row so we can safely
+// overwrite the previous value on each iteration.
+func decodeConstraints(cons applicationConstraints) constraints.Value {
+	res := constraints.Value{}
+
+	// Empty constraints is not an error case, so return early the empty
+	// result.
+	if len(cons) == 0 {
+		return res
+	}
+
+	// Unique spaces, tags and zones:
+	spaces := make(map[string]string)
+	tags := make(map[string]string)
+	zones := make(map[string]string)
+
+	for _, row := range cons {
+		if row.Arch.Valid {
+			res.Arch = &row.Arch.String
+		}
+		if row.CPUCores.Valid {
+			cpuCores := uint64(row.CPUCores.Int64)
+			res.CpuCores = &cpuCores
+		}
+		if row.CPUPower.Valid {
+			cpuPower := uint64(row.CPUPower.Int64)
+			res.CpuPower = &cpuPower
+		}
+		if row.Mem.Valid {
+			mem := uint64(row.Mem.Int64)
+			res.Mem = &mem
+		}
+		if row.RootDisk.Valid {
+			rootDisk := uint64(row.RootDisk.Int64)
+			res.RootDisk = &rootDisk
+		}
+		if row.RootDiskSource.Valid {
+			res.RootDiskSource = &row.RootDiskSource.String
+		}
+		if row.InstanceRole.Valid {
+			res.InstanceRole = &row.InstanceRole.String
+		}
+		if row.InstanceType.Valid {
+			res.InstanceType = &row.InstanceType.String
+		}
+		if row.ContainerType.Valid {
+			containerType := instance.ContainerType(row.ContainerType.String)
+			res.Container = &containerType
+		}
+		if row.VirtType.Valid {
+			res.VirtType = &row.VirtType.String
+		}
+		if row.AllocatePublicIP.Valid {
+			res.AllocatePublicIP = &row.AllocatePublicIP.Bool
+		}
+		if row.ImageID.Valid {
+			res.ImageID = &row.ImageID.String
+		}
+		if row.Space.Valid {
+			spaces[row.Space.String] = row.Space.String
+		}
+		if row.Tag.Valid {
+			tags[row.Tag.String] = row.Tag.String
+		}
+		if row.Zone.Valid {
+			zones[row.Zone.String] = row.Zone.String
+		}
+	}
+
+	// Add the unique spaces, tags and zones to the result:
+	if len(spaces) > 0 {
+		spacesSlice := make([]string, 0, len(spaces))
+		for _, space := range spaces {
+			spacesSlice = append(spacesSlice, space)
+		}
+		res.Spaces = &spacesSlice
+	}
+	if len(tags) > 0 {
+		tagsSlice := make([]string, 0, len(tags))
+		for _, tag := range tags {
+			tagsSlice = append(tagsSlice, tag)
+		}
+		res.Tags = &tagsSlice
+	}
+	if len(zones) > 0 {
+		zonesSlice := make([]string, 0, len(zones))
+		for _, zone := range zones {
+			zonesSlice = append(zonesSlice, zone)
+		}
+		res.Zones = &zonesSlice
+	}
```

Comment:
knit: It would increase readability if, for example, `spaces` was a `set.Strings` and you called `spaces.Values()` or `spaces.SortedValues()`